### PR TITLE
[mqtt.homeassistant] Implement template schema lights

### PR DESCRIPTION
**Pull request by ccutrer**

### CODE REVIEW

**Comment by ccutrer** [comment]
Yay! The initial commit of this PR only happened nearly 2 years ago. And now it can actually be merged in!

**Comment by lsiepel** [comment]
> Yay! The initial commit of this PR only happened nearly 2 years ago. And now it can actually be merged in!

What took you so long ;-)

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+        publishState(binding, template);
+    }
+
+    private boolean handleColorTempCommand(Command command) {
+        if (command instanceof DecimalType) {
+            command = new QuantityType<>(((DecimalType) command).toBigDecimal(), Units.MIRED);
+        }
+        if (command instanceof QuantityType) {
+            QuantityType<?> mireds = ((QuantityType<?>) command).toInvertibleUnit(Units.MIRED);
+            if (mireds == null) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java:149]:
```suggestion
        if (command instanceof QuantityType<?> quntityTypeCommand) {
            QuantityType<?> mireds = quntityTypeCommand.toInvertibleUnit(Units.MIRED);
```

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+        ChannelStateUpdateListener listener = this.channelStateUpdateListener;
+
+        String value;
+
+        String template = channelConfiguration.stateTemplate;
+        if (template != null) {
+            value = transform(template, state.toString());
+            if (value == null || value.isEmpty()) {
+                onOffValue.update(UnDefType.NULL);
+            } else if (value.equals("on")) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java:201]:
```suggestion
            } else if ("on".equals(value)) {
```

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+                onOffValue.update(OnOffType.ON);
+            } else if (value.equals("off")) {
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java:203]:
```suggestion
            } else if ("off".equals(value)) {
```

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        if (template != null) {
+            Integer brightness = getColorChannelValue(template, state.toString());
+            if (brightness == null) {
+                brightnessValue.update(UnDefType.NULL);
+                colorValue.update(UnDefType.NULL);
+            } else {
+                brightnessValue.update((PercentType) brightnessValue.parseMessage(new DecimalType(brightness)));
+                if (colorValue.getChannelState() instanceof HSBType) {
+                    HSBType color = (HSBType) colorValue.getChannelState();
+                    colorValue.update(new HSBType(color.getHue(), color.getSaturation(),
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java:230]:
```suggestion
                if (colorValue.getChannelState() instanceof HSBType color) {
                    colorValue.update(new HSBType(color.getHue(), color.getSaturation(),
```

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        } catch (NumberFormatException e) {
+            logger.warn("Applying template {} for component {} failed: {}", template, getHaID().toShortTopic(),
+                    e.getMessage());
+            return null;
+        }
+    }
+
+    private @Nullable String transform(String template, Map<String, @Nullable Object> binding) {
+        Optional<String> result = transformation.apply(template, binding);
+        return result.orElse(null);
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java:306]:
```suggestion
        return transformation.apply(template, binding).orElse(null);
```

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    private @Nullable String transform(String template, String value) {
+        Optional<String> result = transformation.apply(template, value);
+        return result.orElse(null);
+    }
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java:312]:
```suggestion
        return transformation.apply(template, value).orElse(null);
```

File: bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+        } else {
+            template = Objects.requireNonNull(channelConfiguration.commandOnTemplate);
+            binding.put(TemplateVariables.STATE, "on");
+            if (channelConfiguration.brightnessTemplate != null) {
+                binding.put(TemplateVariables.BRIGHTNESS,
+                        state.getBrightness().toBigDecimal().multiply(factor).intValue());
+            }
+            if (hasColorChannel) {
+                PercentType[] rgb = state.toRGB();
+                binding.put(TemplateVariables.RED, rgb[0].toBigDecimal().multiply(factor).intValue());
```
Comment (lsiepel) [inline bundles/org.openhab.binding.mqtt.homeassistant/src/main/java/org/openhab/binding/mqtt/homeassistant/internal/component/TemplateSchemaLight.java:132]:
```suggestion
                PercentType[] rgb = ColorUtil.hsbToRgbPercent(state);
```

**Review by lsiepel — CHANGES_REQUESTED** [review]
Some code style improvements.

**Review by lsiepel — APPROVED** [review]
Thanks, LGTM

**Pull request merged**